storage-proofs/post/src/fallback/vanilla.rs

```diff
@@ -355,6 +356,84 @@ impl<'a, Tree: 'a + MerkleTreeTrait> ProofScheme<'a> for FallbackPoSt<'a, Tree>
         Ok(partition_proofs)
     }
 
+    fn prove_all_partitions_with_challenges<'b>(
+        pub_params: &'b Self::PublicParams,
+        pub_inputs: &'b Self::PublicInputs,
+        priv_inputs: &'b Self::PrivateInputs,
+        partition_count: usize,
+        sector_challenges: &BTreeMap<SectorId, Vec<u64>>,
+    ) -> Result<Vec<Self::Proof>> {
```

Comment:
I think maybe this is not decomposed right or fully enough. I'm not sure partitions and challenges should be mixed like this. There should be two distinct actions:

- Generate vanillla proofs for a single sector, given the challenges for that sector.
- Group vanilla proofs into partitions.

When decomposed as planned, the vanilla proofs will have been generated separately, but not grouped into partitions.  I think we will need a `partition_vanilla_proofs` (note 'partition' is a verb here) function which just takes all the vanilla proofs and creates partitions proofs from them. (It happens to be the case that both Window and Winning PoSt only use a single partition, but we don't need to and shouldn't rely on that.) This will include the padding logic from `prove_all_partitions` — for the case that fewer than the 'required' number of vanilla proofs are provided.